text/3541-async-drop.md

```diff
@@ -0,0 +1,222 @@
+    sleep(Duration::from_secs(2)).await;
+    println!("Main function completed");
+}
+```
+
+Moreover, to seamlessly integrate native async drop support into Rust, it is essential to outline how executors should handle asynchronous resource cleanup. This section also provides guidelines for developers using executors to ensure consistent and correct execution of async drop logic.
+
+1. **Recognition of AsyncDrop Trait**
+   Executors, such as `MyExecutor` in the following example, should be updated to recognize types implementing the `AsyncDrop` trait. This can be achieved through a mechanism such as trait bounds or associated types to identify futures that require special handling for asynchronous cleanup.
+
+   ```rust
+   trait Executor {
+       fn spawn(&self, future: impl Future);
+   }
+
+   impl<T: AsyncDrop> Executor for MyExecutor {
+       // Your custom cleanup logic goes here
+   }
+   ```
+
+   Here, `T: AsyncDrop` indicates that this executor can work with futures (`T`) that implement the AsyncDrop trait. The `AsyncDrop` trait serves as a marker, signaling that the associated type requires special handling for asynchronous cleanup.
+
+1. **Async Drop Invocation**
+   Executors must ensure that the async `drop` method is invoked when a future implementing the `AsyncDrop` trait goes out of scope or completes. This involves tracking the lifecycle of futures and, upon termination, triggering the async drop logic before releasing associated resources.
+
+   ```rust
+   async fn execute_future<T: AsyncDrop>(&self, fut: T) {
+       // Execute the future
+       let result = block_on(fut);
+   
+       // Perform async drop before releasing resources
+       fut.drop().await;
+   }
+   ```
+
+   The `futures::executor::block_on` function is employed to execute the future (`fut`), pausing the current thread until the future completes and produces a result. Following the completion of the future's execution, the executor invokes the `drop` method on the future (`fut`). This ensures the execution of asynchronous drop logic, allowing the future to perform any necessary cleanup operations before resources are released.
+
+1. **Context Propagation**
+   To maintain a consistent execution context for async drops, executors should propagate the appropriate `Waker` and `Context` information to the async drop method. This ensures that async drops can interact with the executor environment and make executor-specific decisions during cleanup.
+
+   ```rust
+   async fn execute_future<T: AsyncDrop + Future + Copy>(&self, fut: T) {
+       // Set up the execution context
+       let waker = Waker::noop();
+       let mut cx = Context::from_waker(&waker);
+       
+       // Execute the future with the provided context
+       let mut pinned = std::pin::pin!(fut);
+       let _ = pinned.as_mut().poll(&mut cx); 
+       
+       // Perform async drop within the same context
+       fut.drop().await;
+   }
+   ```
```

Comment:
A noop waker is nice to make the simplier busyloop executor, but is definitely not a good example when you want to show how to handle `Waker`s.

`fut.drop()` is supposed to take an `&mut self` according to your RFC, but here `fut` was pinned so it is a `Pin<&mut T>` which you can't get a `&mut T` from because it doesn't necessarily implement `Unpin` (and you **don't want** to require `Unpin` for every possible `Future`).

Also, why does it even require `T: Copy`?